[Filter] Add 'disabled' prop

PabloVallejo opened the pull request:
<!--
  ☝️How to write a good PR title:
  - Prefix it with [ComponentName] (if applicable), for example: [Button]
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
  - Prefix it with [WIP] while it’s a work in progress
-->

### WHY are these changes introduced?

Related to https://github.com/Shopify/store/issues/10681

As part of the solution for https://github.com/Shopify/store/issues/10681 we need to disable the `Filters` in the Products Index Page. Currently the filter component doesn't have a prop to disable it.

### WHAT is this pull request doing?

This PR adds a `disabled` prop to the `Filter` component.

When the `Filter` is disabled:

* Search-bar is disabled
* Actions from `ConnectedFilterControl` are disabled.
* Popover is closed 
* Sheet filter togglers are disabled
* Sheet filter titles are subdued


| Disabled filters        | Disabled sheet     | 
| ------------- |:-------------:| 
| ![image](https://user-images.githubusercontent.com/1751074/67976286-2bf76d00-fbec-11e9-84b2-8ce5a151445d.png) | ![image](https://user-images.githubusercontent.com/1751074/67976233-1a15ca00-fbec-11e9-837c-dc5f19f192f0.png) | 

<!--
  Summary of the changes committed.

  Before / after screenshots are appreciated for UI changes. Make sure to include alt text that describes the screenshot.

  If you include an animated gif showing your change, wrapping it in a details tag is recommended. Gifs usually autoplay, which can cause accessibility issues for people reviewing your PR:

    <details>
      <summary>Summary of your gif(s)</summary>
      <img src="..." alt="Description of what the gif shows">
    </details>
-->

## <!-- ℹ️ Delete the following for small / trivial changes -->

### How to 🎩

* Paste the playground code and run the project 
* From the playground, see that all the filters are active
* ✅Click the account status button and select any option: See that the filters get disabled
* ✅Click `More filters` and from the Account status filter select any option: See that the main filters get disabled and the filters in the Sidebar get their title subdued!

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<!--
  Give as much information as needed to experiment with the component
  in the playground.
-->

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React, {useState, useCallback} from 'react';
import {
  Page,
  TextField,
  Avatar,
  Filters,
  Card,
  TextStyle,
  ResourceList,
  ChoiceList,
} from '../src';

function ResourceListFiltersExample() {
  const [accountStatus, setAccountStatus] = useState(null);
  const [moneySpent, setMoneySpent] = useState(null);
  const [taggedWith, setTaggedWith] = useState(null);
  const [queryValue, setQueryValue] = useState(null);

  const handleAccountStatusChange = useCallback(
    (value) => setAccountStatus(value),
    [],
  );
  const handleMoneySpentChange = useCallback(
    (value) => setMoneySpent(value),
    [],
  );
  const handleTaggedWithChange = useCallback(
    (value) => setTaggedWith(value),
    [],
  );
  const handleFiltersQueryChange = useCallback(
    (value) => setQueryValue(value),
    [],
  );
  const handleAccountStatusRemove = useCallback(
    () => setAccountStatus(null),
    [],
  );
  const handleMoneySpentRemove = useCallback(() => setMoneySpent(null), []);
  const handleTaggedWithRemove = useCallback(() => setTaggedWith(null), []);
  const handleQueryValueRemove = useCallback(() => setQueryValue(null), []);
  const handleFiltersClearAll = useCallback(() => {
    handleAccountStatusRemove();
    handleMoneySpentRemove();
    handleTaggedWithRemove();
    handleQueryValueRemove();
  }, [
    handleAccountStatusRemove,
    handleMoneySpentRemove,
    handleQueryValueRemove,
    handleTaggedWithRemove,
  ]);

  const filters = [
    {
      key: 'accountStatus',
      label: 'Account status',
      filter: (
        <ChoiceList
          title="Account status"
          titleHidden
          choices={[
            {label: 'Enabled', value: 'enabled'},
            {label: 'Not invited', value: 'not invited'},
            {label: 'Invited', value: 'invited'},
            {label: 'Declined', value: 'declined'},
          ]}
          selected={accountStatus || []}
          onChange={handleAccountStatusChange}
          allowMultiple
        />
      ),
      shortcut: true,
    },
    {
      key: 'taggedWith',
      label: 'Tagged with',
      filter: (
        <TextField
          label="Tagged with"
          value={taggedWith}
          onChange={handleTaggedWithChange}
          labelHidden
        />
      ),
      shortcut: true,
    },
  ];

  const appliedFilters = [];
  if (!isEmpty(accountStatus)) {
    const key = 'accountStatus';
    appliedFilters.push({
      key,
      label: disambiguateLabel(key, accountStatus),
      onRemove: handleAccountStatusRemove,
    });
  }
  if (!isEmpty(moneySpent)) {
    const key = 'moneySpent';
    appliedFilters.push({
      key,
      label: disambiguateLabel(key, moneySpent),
      onRemove: handleMoneySpentRemove,
    });
  }
  if (!isEmpty(taggedWith)) {
    const key = 'taggedWith';
    appliedFilters.push({
      key,
      label: disambiguateLabel(key, taggedWith),
      onRemove: handleTaggedWithRemove,
    });
  }

  const isDisabled =
    appliedFilters.length > 0 && appliedFilters[0].key === 'accountStatus';

  return (
    <div style={{height: '568px'}}>
      <Card>
        <ResourceList
          resourceName={{singular: 'customer', plural: 'customers'}}
          filterControl={
            <Filters
              queryValue={queryValue}
              filters={filters}
              appliedFilters={appliedFilters}
              onQueryChange={handleFiltersQueryChange}
              onQueryClear={handleQueryValueRemove}
              onClearAll={handleFiltersClearAll}
              disabled={isDisabled}
            />
          }
          items={[
            {
              id: 341,
              url: 'customers/341',
              name: 'Mae Jemison',
              location: 'Decatur, USA',
            },
            {
              id: 256,
              url: 'customers/256',
              name: 'Ellen Ochoa',
              location: 'Los Angeles, USA',
            },
          ]}
          renderItem={(item) => {
            const {id, url, name, location} = item;
            const media = <Avatar customer size="medium" name={name} />;

            return (
              <ResourceList.Item
                id={id}
                url={url}
                media={media}
                accessibilityLabel={`View details for ${name}`}
              >
                <h3>
                  <TextStyle variation="strong">{name}</TextStyle>
                </h3>
                <div>{location}</div>
              </ResourceList.Item>
            );
          }}
        />
      </Card>
    </div>
  );

  function disambiguateLabel(key, value) {
    switch (key) {
      case 'moneySpent':
        return `Money spent is between $${value[0]} and $${value[1]}`;
      case 'taggedWith':
        return `Tagged with ${value}`;
      case 'accountStatus':
        return value.map((val) => `Customer ${val}`).join(', ');
      default:
        return value;
    }
  }

  function isEmpty(value) {
    if (Array.isArray(value)) {
      return value.length === 0;
    } else {
      return value === '' || value == null;
    }
  }
}

export function Playground() {
  return (
    <Page title="Playground">
      <ResourceListFiltersExample />
    </Page>
  );
}

```

</details>

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [x] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the [Polaris UI kit](https://polaris.shopify.com/resources/polaris-ui-kit)

<!--
  When tophatting, please check the UNRELEASED entry for consistency, as per the [guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md).
-->

<!--
  If you don't have access to Percy, please request access in the #polaris Slack channel.
-->

dleroux commented on src/components/Filters/Filters.tsx:
```diff
@@ -171,9 +179,14 @@ class Filters extends React.Component<ComposedProps, State> {
             <div className={styles.FilterTriggerLabelContainer}>
-              <h2 className={styles.FilterTriggerTitle}>{filter.label}</h2>
+              <h2 className={styles.FilterTriggerTitle}>
+                <TextStyle {...filterTriggerTitleTextStyle}>
```
This would be a little less code
`<TextStyle  variation={disabled ? VariationValue.Subdued : undefined}>`

dleroux commented on src/components/Filters/README.md:
````diff
@@ -650,3 +650,130 @@ function FiltersExample() {
 }
 ```
+
+### Disabling filters
````
Disabled filters

dleroux commented:
Code looks good. Thank you for adding the examples!

One thing... The tags that get generated should probably also be disabled as well because I assume removing them would ### have a negative impact on the filtering in your case.

PabloVallejo commented:
The way our workflow works is that filters are enabled and once merchants add one particular filter, they get disabled and we should a text that they need to remove that filter in order to use the other ones. 

For this case, having the tags available allows merchants to remove the filter and be able to add more again.

I guess we could disable everything and provide the merchants with an action link/button to remove applied filters enabling them again but removing the tags seems to be clean for them.  We do that now in Rails.

I strongly prefer this solution, but if you'd rather have everything disabled I can add the changes too.

dleroux commented:
I see, a bit of a special usecase. I wonder if someone involved in the dev/design of this component can chime in. @sarahill, @danrosenthal

dleroux commented:
I can see someone down the road thinking this isn't intended and breaking your use case.

PabloVallejo commented:
@dleroux analyzing our use case, we decided that along with the descriptive text on why the filters get disabled when you apply a particular filter, we're going to add a link to reset the filters. 

In that sense, the filters components will be fully disabled once you pass `disabled` prop as `true`. Keeping consistency between the name of the prop and the state of the component.

dleroux commented:
👍 I think from an accessibility perspective it might be a better choice/ux. Thanks, Pablo.

PabloVallejo commented:
@dleroux sounds good! Yea, in that way the disabled prop ends up being more semantic!

The change is up and just fixed the merge conflict ✅

devonpmack commented on src/components/Filters/components/ConnectedFilterControl/ConnectedFilterControl.tsx:
```diff
@@ -183,30 +184,34 @@ export class ConnectedFilterControl extends React.Component<
     );
-  });
-}
+  }
```
Is this to stop them from clicking on popover elements if the filter gets disabled while the popover is already open? If so could we just close the popover if the filter is disabled instead?

devonpmack commented on src/components/Filters/tests/Filters.test.tsx:
```diff
@@ -293,6 +293,84 @@ describe('<Filters />', () => {
   });
+
+  describe('disabled prop is passed as true', () => {
+    it('filters search field is diabled', () => {
```
```suggestion
    it('disables filter search field', () => {
```

devonpmack commented on src/components/Filters/tests/Filters.test.tsx:
```diff
@@ -293,6 +293,84 @@ describe('<Filters />', () => {
+      expect(resourceFilters.find(TextField).prop('disabled')).toBe(true);
+    });
+
+    it('sheet toggle button is disabled', () => {
```
```suggestion
    it('disables sheet toggle button', () => {
```

devonpmack commented on src/components/Filters/tests/Filters.test.tsx:
```diff
@@ -293,6 +293,84 @@ describe('<Filters />', () => {
+      ).toBe(true);
+    });
+
+    it('sheet filter toggle button is disabled', () => {
```
```suggestion
    it('disables sheet filter toggle buttons', () => {
```

devonpmack commented on src/components/Filters/tests/Filters.test.tsx:
```diff
@@ -293,6 +293,84 @@ describe('<Filters />', () => {
+      });
+    });
+
+    it('sheet filters heading text is subdued', () => {
```
```suggestion
    it('subdues sheet filter heading text', () => {
```

devonpmack commented on src/components/Filters/components/ConnectedFilterControl/ConnectedFilterControl.tsx:
```diff
@@ -183,30 +184,34 @@ export class ConnectedFilterControl extends React.Component<
     );
-  });
-}
+  }
```
Is this to stop them from clicking on popover elements if the filter gets disabled while the popover is already open? If so could we just close the popover if the filter is disabled instead?

chloerice replied:
Ahh, I think the diff made this confusing. This was a functional component but Pablo converted it to a class because these two functions both need access to `props.disabled`.

devonpmack replied:
Oops

devonpmack commented:
When tophatting, it seems like I can't remove the filter after enabling it

![image](https://user-images.githubusercontent.com/21130966/68132721-2b611e00-feed-11e9-95e4-e23e2816c453.png)

devonpmack commented:
![image](https://user-images.githubusercontent.com/21130966/68132852-62cfca80-feed-11e9-8acc-973036d2dc51.png)

It would be great if we could also disable the onHover of the drop downs. Right now it changes your cursor to 'link mode' and darkens the button. This makes it seem like you can click it even though you can't. It doesn't really match the way other disabled buttons work in polaris. I understand that this could be hard to fix though.

chloerice commented on src/components/Filters/Filters.tsx:
```diff
@@ -75,6 +75,8 @@ export interface FiltersProps {
   onQueryFocus?(): void;
   /** The content to display inline with the controls */
   children?: React.ReactNode;
+  /** Disables the filters when set to true */
```
```suggestion
  /** Disable all filters */
```

chloerice commented on src/components/Filters/README.md:
```diff
@@ -650,3 +650,130 @@ function FiltersExample() {
+              disabled
+            >
+              <div style={{paddingLeft: '8px'}}>
+                <Button onClick={() => console.log('New filter saved')}>
```
Let's disable the save button too since the filters themselves are disabled.

```suggestion
                <Button disabled onClick={() => console.log('New filter saved')}>
```

dleroux commented on src/components/Filters/Filters.tsx:
```diff
@@ -171,9 +174,16 @@ class Filters extends React.Component<ComposedProps, State> {
-              <h2 className={styles.FilterTriggerTitle}>{filter.label}</h2>
+              <h2 className={styles.FilterTriggerTitle}>
+                <TextStyle
+                  variation={disabled ? VariationValue.Subdued : undefined}
```
Is there a use-case for disabling individual filters?

Good catch on it not being needed since the sheet can't be opened.

chloerice replied:
I think so, but if you think it's out of scope for this PR it can just be added when it comes up in the future.

PabloVallejo replied:
To give some context, our workflow is:

* While all filters are active
* Merchant opens the sheet
* Sets a filter by collection: This action disables other filters but the sheet is still open. 

So we'll need to disable filters in the sheet, but we don't need to disable all of them, so adding support for disabling individual filters is the way to go.

dleroux commented:
> It would be great if we could also disable the onHover of the drop downs. Right now it changes your cursor to 'link mode' and darkens the button. This makes it seem like you can click it even though you can't. It doesn't really match the way other disabled buttons work in polaris. I understand that this could be hard to fix though.

Hi @devonpmack are you saying with the filters disabled you were able to open the sheet? These should be clickable if filters aren't disabled, they work as an accordion. Is that no the case for you?

PabloVallejo commented on src/components/Filters/README.md:
````diff
@@ -650,3 +650,130 @@ function FiltersExample() {
 }
 ```
+
+### Disabled filters
````
Added the example!

PabloVallejo commented:
@chloerice @dleroux @devonpmack , I just pushed some code that:

* Adds support for disabling individual filters
* Adds a new prop: `HelpText` to Filters & add tests for it
* Fix filter tests

I was chatting IRL with Devon yesterday and he noticed that not being able to remove a filter feels blocking. To add some context, here's the workflow we're trying to replicate:

![filters-gif](https://user-images.githubusercontent.com/1751074/68215939-15fff880-ffae-11e9-9e41-47b98af92e22.gif)

We have two options to do this:

### 1. Filter tags are available for merchants to remove them

![image](https://user-images.githubusercontent.com/1751074/68215161-c3720c80-ffac-11e9-9acd-ead58339c534.png)


### 2. Filter tags are not available and merchants are given a help text with a link to reset them and make them available.

![image](https://user-images.githubusercontent.com/1751074/68216236-97578b00-ffae-11e9-9bd9-5ef325f4774c.png)

Allowing the removal of tags feels more natural in my opinion but I'm open to both options.

PabloVallejo commented:
@dleroux what @devonpmack meant is that if before disabling the filters you open the sheet, and then disable the filters, you'd see the headers as links.

The sheet can be open before the filters are disabled.

dleroux commented:
I see what you are saying about leaving the tags enabled. Maybe a designer can weigh in, @sarahill ?

I was thinking if you disable because a request is in progress, removing the tag would defy the purpose. In this case I can see why you would want it.

PabloVallejo commented:
@dleroux sounds good! I'm going to talk to @sarahill to discuss this one! Thanks 🙏🏼🙏🏼

sarahill commented:
Hey @PabloVallejo, I've got a few questions so I have some more context:

- What's the reasoning for only 1 filter at a time?
- Does the search count as a filter (could we leave the search enabled)?

PabloVallejo commented:
👋🏼 @sarahill 

The TL;DR is that because of backend limitation, the collection filter cannot be combined with other filters, therefore:

* We're limiting merchants to add other filters if they filter by collection.
* The search does count as a filter and cannot be combined with the collections filter hence cannot be enabled.

chloerice commented on src/components/Filters/Filters.tsx:
```diff
@@ -212,6 +233,7 @@ class Filters extends React.Component<ComposedProps, State> {
         rightPopoverableActions={this.transformFilters(filters)}
```
The `PopoverableAction` interface defined in `ConnectedFilterControl.tsx` also needs to be updated to extend `DisableableAction` instead of `Action`.

chloerice replied:
The type of the `action` argument of the `activatorButtonFrom` method needs to be updated to `PopoverableAction` instead of `BaseAction`.

PabloVallejo replied:
Nice suggestion, this allows for disabling individual activators! Added it to the code!

chloerice commented on src/components/Filters/components/ConnectedFilterControl/ConnectedFilterControl.tsx:
```diff
@@ -183,30 +184,34 @@ export class ConnectedFilterControl extends React.Component<
+      return (
+        <Item key={action.key}>
+          <Popover
+            active={this.props.disabled ? false : action.popoverOpen}
```
```suggestion
            active={action.popoverOpen}
```

I don't think this needs to be conditional since the popover activator is `disabled` when the filter is `disabled`.

PabloVallejo commented:
Discussing IRL with @dleroux and @sarahill the approach that made the most sense was approach number 2: Disabling filters and tags altogether using the disable prop and giving merchants a link to reset the filters:

![](https://user-images.githubusercontent.com/1751074/68216236-97578b00-ffae-11e9-9bd9-5ef325f4774c.png)

I've updated the code accordingly and addressed review comments.

I also did a couple of tests with extracting `Filter` component from the  `ResourceList` but a refactor would be needed in order to allow this as the styles get messed up (see screenshot). So I'll keep things inside of `ResourceList`: 

![image](https://user-images.githubusercontent.com/1751074/68302735-938c3d00-0070-11ea-880c-c61c228f04bf.png)


Similarly, I tried adding the `HelpText` after the filters but was not happy with the looks, what do you think @dleroux, @sarahill:

![image](https://user-images.githubusercontent.com/1751074/68302925-e4039a80-0070-11ea-87b4-c149089aebef.png)

Deciding on the placement of the help text would be the last detail and this PR should be good to 🚢

sarahill commented:
@pablogolan cc: @daniedleroux 

For the text, can we do something like the attached? If it's easier to leave the text below the tags that's ok. I think this reads a little better and then on small screens the text + link can go below.

![Screen Shot 2019-11-06 at 10 59 17 AM](https://user-images.githubusercontent.com/8629173/68314673-8f6a1a80-0084-11ea-8d16-9b50477f232b.png)

dleroux commented:
@sarahill, yes that would be ideal and in line with what we do in the bulk actions today. However not possible without a new prop. Is this a prop we see value in? an `inlineAction` maybe?

If not then it might be best to merge the disable stuff and then fork it for your use case @PabloVallejo and then delete it when the limitation is no longer.

pablogolan commented:
@sarahill I think you meant to ping Pablo Vallejo, not me.

pablogolan commented:
Can someone please remove me from here. I am still getting emails regarding this issue. Thanks!

PabloVallejo commented:
Hey Pablo, sorry about that, you can unsubscribe:

![image](https://user-images.githubusercontent.com/1751074/68395400-a6207800-013d-11ea-83b2-bec1189b2856.png)


But since the PR was just merged we won't be discussing on it anymore! 👌🏼👍🏼
